### DISCUSSION TITLE
How it works with TPM?

### DISCUSSION THREAD

**Discussion by TingALin** [tearust/teaproject#135] [General]
Hi again,

I am having hard time to understand the abstraction on "putting the result on chain after the computation so it will be useless for the hacker to do the hacking" from the TPM function, do you have an example/ a link to help me understand this concept from Tea Project's perspective?

Thank you:)

**Comment by kevingzhang** [comment]
Can you point out the link to your quote "putting the result on chain after the computation so it will be useless for the hacker to do the hacking". 

Without the context, I can guess the meaning of that sentence. It means the computation happened inside enclave which is supposed to very hard to hack. So once the result is confirmed, it is safe to store on chain, and the chain (layer1) is supposed to be tamper-proof storage. 

Once I have the context, I may update my answer.

**Reply by TingALin** [reply]
Yes, you're right on what I'm asking, and I phrased this sentence by watching some of your Youtube videos(maybe some writings somewhere else) in Chinese actually, didn't remember what exactly the statement is but now I can fine one of the source here: https://www.youtube.com/watch?v=4Gaeg-3pcjE at 7:36.

Actually, from your answer if there is a example of what is the result and why none can know the result before it's on chain would be good enough, or is it a concept of TPM(if so, any links for good SIMPLE explanation)?

**Reply by kevingzhang** [reply]
I think you are asking how "enclave" work. Enclave was designed to run code inside a hardware protected isolated environment. That means anyone outside the enclave knows nothing about what happened inside the enclave. Consider the concept of "blackhole".

Until the result is release to the chain outside of the enclave and uploaded to the chain, no one knows the result.

AWS Nitro is one type of enclave that we can rent. For more about AWS Nitro enclave, please go to AWS website

**Comment by kevingzhang** [comment]
> Yes, you're right on what I'm asking, and I phrased this sentence by watching some of your Youtube videos(maybe some writings somewhere else) in Chinese actually, didn't remember what exactly the statement is but now I can fine one of the source here: https://www.youtube.com/watch?v=4Gaeg-3pcjE at 7:36.
> 
> Actually, from your answer if there is a example of what is the result and why none can know the result before it's on chain would be good enough, or is it a concept of TPM(if so, any links for good SIMPLE explanation)?

**Comment by kevingzhang** [comment]
BTW, TPM is not enclave. TPM doesn't run any customer's code. It is enclave that execute the code.

TPM is used to protect the enclave and make the enclave "verifiable"
